crates/findex/src/implementations/redis/findex.rs

```diff
@@ -0,0 +1,434 @@
+        master_key: &[u8; 16],
+        label: &[u8],
+        keywords: HashSet<Keyword>,
+    ) -> Result<HashMap<Keyword, HashSet<Location>>, FindexRedisError> {
+        let res = FindexSearch::search(
+            self,
+            &KeyingMaterial::<MASTER_KEY_LENGTH>::from(*master_key),
+            &Label::from(label),
+            keywords,
+        )
+        .await?;
+        Ok(res)
+    }
+
+    /// Replaces all the Index Entry Table UIDs and values. New UIDs are derived
+    /// using the given label and the KMAC key derived from the new master key.
+    /// The values are decrypted using the DEM key derived from the master key
+    /// and re-encrypted using the DEM key derived from the new master key.
+    ///
+    /// Randomly selects index entries and recompact their associated chains.
+    /// Chains indexing no existing location are removed. Others are recomputed
+    /// from a new keying material. This removes unneeded paddings. New UIDs are
+    /// derived for the chain and values are re-encrypted using a DEM key
+    /// derived from the new keying material.
+    ///
+    /// - `master_key`                      : master key used to generate the
+    ///   current index
+    /// - `new_master_key`                  : master key used to generate the
+    ///   new index
+    /// - `label`                           : label used to generate the new
+    ///   index
+    /// - `num_reindexing_before_full_set`  : average number of calls to compact
+    ///   needed to recompute all of the Chain Table.
+    ///
+    /// **WARNING**: the compact operation *cannot* be done concurrently with
+    /// upsert operations. This could result in corrupted indexes.
+    pub async fn compact(
+        &self,
+        master_key: &[u8; 16],
+        new_master_key: &[u8; 16],
+        label: &[u8],
+        num_reindexing_before_full_set: u32,
+    ) -> Result<(), FindexRedisError> {
+        FindexCompact::compact(
+            self,
+            &KeyingMaterial::<MASTER_KEY_LENGTH>::from(*master_key),
+            &KeyingMaterial::<MASTER_KEY_LENGTH>::from(*new_master_key),
+            &Label::from(label),
+            num_reindexing_before_full_set,
+        )
+        .await?;
+        Ok(())
+    }
+}
+
+impl FindexCallbacks<FindexRedisError, UID_LENGTH> for FindexRedis {
+    async fn progress(
+        &self,
+        _results: &HashMap<Keyword, HashSet<IndexedValue>>,
+    ) -> Result<bool, FindexRedisError> {
+        //TODO: allow passing callback fn on connect
+        Ok(true)
+    }
+
+    async fn fetch_all_entry_table_uids(&self) -> Result<Uids<UID_LENGTH>, FindexRedisError> {
+        let keys: Vec<Vec<u8>> = self.mgr.clone().keys(key(FindexTable::Entry, b"*")).await?;
+        trace!("fetch_all_entry_table_uids num keywords: {}", keys.len());
+        Ok(Uids(
+            keys.iter()
+                .map(|v| {
+                    let mut uid = [0u8; UID_LENGTH];
+                    uid.copy_from_slice(&v[4..]);
```

Comment:
The `4` here is a little magic, maybe store `b"fe::"` inside a constant and use the length of this constant instead?

Reply:
Now in a constant 

```
/// The length of the prefix of the table name in bytes
/// 0x00ee for the entry table
/// 0x00ef for the chain table
const TABLE_PREFIX_LENGTH: usize = 2;
```